c_src/main.c

```diff
@@ -569,9 +560,10 @@ erlfdb_future_cancel(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
     enif_mutex_lock(future->lock);
 
     future->cancelled = true;
+    enif_mutex_unlock(future->lock);
+
     fdb_future_cancel(future->future);
 
-    enif_mutex_unlock(future->lock);
```

Comment:
The `ErlFDBFuture* future` is once created and never changed except for the `canceled` field until the destructor is called.  So, there is no need to synchronize the whole future.

In the destructor, we own the future exclusively and call only thread-safe funs.

Reply:
Previously, the call sequence

`erlfdb_future_cancel(...)` -> `enif_mutex_unlock(future->lock)` -> `fdb_future_cancel(future->future)` ->  `erlfdb_future_cb(...)` -> `enif_mutex_unlock(future->lock)` lead to a deadlock.